File: leet_code_algorithms/invert_binary_tree/Invert_binary_tree.java
package invert_binary_tree;

public class Invert_binary_tree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public TreeNode invertTree(TreeNode root) {
        if(root == null) {
            return root;
        }
        TreeNode leftPlaceholder = root.left;
        root.left = root.right;
        root.right = leftPlaceholder;
        invertTree(root.left);
        invertTree(root.right);
        return root;
    }
}

File: leet_code_algorithms/invert_binary_tree/invert_binary_tree.py
# My Solution

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
    def invertNode(node: Optional[TreeNode]):
        current = node
        if current == None:
            return
        leftPlaceholder = current.left if current.left else None
        current.left = node.right if node.right else None
        current.right = leftPlaceholder
        invertNode(current.left)
        invertNode(current.right)
    if root == None:
        return root
    else:
        invertNode(root)
    return root

# Optimal solution
def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
    if root is None:
        return root
    self.invertTree(root.left)
    self.invertTree(root.right)
    root.left, root.right = root.right, root.left
    return root
